src/providers/mod.rs

```diff
@@ -0,0 +1,112 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::{
+  boxed::Box,
+  collections::HashMap,
+  error::Error,
+  fmt::{self, Display, Formatter},
+};
+
+use self::{mock::MockProvider as Mock, ses::SesProvider as Ses};
+use settings::Settings;
+
+mod mock;
+mod ses;
+
+trait Provider
+{
+  fn send(
+    &self,
+    to: String,
+    cc: Vec<String>,
+    subject: String,
+    body_text: String,
+    body_html: Option<String>,
+  ) -> Result<String, ProviderError>;
+}
+
+#[derive(Debug)]
+pub struct ProviderError
+{
+  description: String,
+}
+
+impl ProviderError
+{
+  pub fn new(description: String) -> ProviderError
+  {
+    ProviderError { description }
+  }
+}
+
+impl Error for ProviderError
+{
+  fn description(&self) -> &str
+  {
+    &self.description
+  }
+}
+
+impl Display for ProviderError
+{
+  fn fmt(&self, formatter: &mut Formatter) -> fmt::Result
+  {
+    write!(formatter, "{}", self.description)
+  }
+}
+
+pub struct Providers
+{
+  settings: Settings,
+  providers: HashMap<String, Box<Provider>>,
```

Comment:
And this map is used to direct requests to the appropriate provider, so all the caller needs to know is the provider name ("ses" or "smtp" or whatever).

Reply:
I've only read a little bit about enums and haven't used them in anger yet, so I'm not quite sure what that implementation would look like in detail.

It seems like we'd still need something in place of the `HashMap` here to do the mapping from the string to the enum value anyway. And I don't know what those enum values would look like. Would we still have the `Provider` trait and different implementations thereof, or would they directly be variants of the enum? Can functions be enum values? (not looking for answers here, just listing some of the things I don't understand yet)

I can play around with it tomorrow though and see where I end up.

Reply:
> I can play around with it tomorrow though and see where I end up.

On the other hand, it's probably very easy for us to get lost in the type-system weeds here, I'm very happy for it to stay as a `HashMap` while we focus on fleshing out all the required behaviors for this repo.